blog/2021-05-19-monero-transaction.md

```diff
@@ -0,0 +1,280 @@
+
+There are too many instances of this to document here, so we instead invite the interested reader to check out our [Rust implementation of Monero's bulletproof](https://github.com/comit-network/monero-rs/blob/b41ae9c32cb435a0f0c8e03f450dd248507c43b5/src/bulletproof/mod.rs).
+It is a fork of [`dalek-cryptography/bulletproofs`](https://github.com/dalek-cryptography/bulletproofs) and currently resides in our fork of `monero-rs`, which we will attempt to merge back into the original repository in the future.
+
+####### Bulletproof generators
+
+The algorithm requires the generation of `2 * m * n` elliptic curve generators, where `m` is the number of amounts to hide in a single proof and `n` the maximum number of bits for the amount.
+Monero sets `m = 16` and `n = 64`, for a total of `2 * 1024 = 2048` generators.
+The number of generators actually used depends on the number of outputs in the transaction.
+
+These generators are deterministically computed and are always the same for any bulletproof to be produced.
+The algorithm to generate them looks like this
+
+```
+for (i=0; i<n*m; i++) {
+    Gi[i] = H_p(H_compressed || "bulletproof" || VarInt(i*2+1));
+    Hi[i] = H_p(H_compressed || "bulletproof" || VarInt(i*2));
+}
+```
+
+where `Gi` and `Hi` are the arrays of points where the 2048 generators will be stored; `H_p` is the same Monero hash function used for key image derivation; `H_compressed` is the compressed representation of Monero's `H` generator; and `VarInt` is a function which maps an integer to a VarInt as described [here](https://developers.google.com/protocol-buffers/docs/encoding#varints).
+
+##### Pseudo output commitments
+
+Given that the real inputs of the transaction are hidden within their respective rings, it is not immediately obvious how we can go about proving that the transaction doesn't create money out of thin air.
+
+If we could use the real inputs' Pedersen commitments (without revealing the real inputs themselves), we could actually leverage the mathematical properties of Pedersen commitments to prove that our transaction is not inflationary.
+For example, for a transaction with 1 input and 1 output with commitments `y_0 * G + 100 * H` and `y_1 * G + 90 * H` respectively, and an explicit fee of `10`, a verifier could create a commitment for the fee without a blinder `0 * G + 10 * H` and sum all these `y_0 * G + 100 * H - (y_1 * G + 90 * H) - (0 * G + 10 * H)` resulting in `(y_0 - y_1) * G + 0 * H`, where all the amounts cancel out.
+
+Since this is not the case, we can instead construct a pseudo output commitment per input and actually choose the blinding factors in such a way that they cancel out when combined with the output commitments.
```

Comment:
Nit: If we remove the can, it is more obvious that this is what actually happens.

```suggestion
Since this is not the case, we instead construct a pseudo output commitment per input and actually choose the blinding factors in such a way that they cancel out when combined with the output commitments.
```